### DISCUSSION TITLE
Drive redirection not working in Alt Linux

### DISCUSSION THREAD

**Discussion by Mogaba** [FreeRDP/FreeRDP#9501] [Q&A]
Hello,

when using `freerdp` to connect to `xrdp`  in Alt Linux I can't make drive redirection work:

`xfreerdp /cert:ignore -sec-nla -menu-anims +fonts /kbd:0x00000409 /network:auto /audio /drive:home,/home/mva`

Here's the output:

```
14:55:25:297] [48256:48257] [INFO][com.freerdp.client.x11] - No user name set. - Using login name: mva
[14:55:26:418] [48256:48257] [ERROR][com.winpr.timezone] - Unable to get current timezone rule
[14:55:26:719] [48256:48257] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[14:55:26:719] [48256:48257] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[14:55:26:761] [48256:48269] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service drive [home] (static)
[14:55:26:766] [48256:48257] [INFO][com.freerdp.channels.rdpsnd.client] - [static] Loaded pulse backend for rdpsnd
[14:55:26:766] [48256:48257] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel rdpsnd
[14:55:26:766] [48256:48257] [INFO][com.freerdp.channels.drdynvc.client] - Loading Dynamic Virtual Channel rdpgfx
[14:55:33:416] [48256:48269] [ERROR][com.freerdp.channels.rdpdr.client] - channel [RDPDR] received 21836, expected state RDPDR_CHANNEL_STATE_READY but have state RDPDR_CHANNEL_STATE_NAME_REQUEST, aborting.
[14:55:33:516] [48256:48269] [ERROR][com.freerdp.channels.rdpdr.client] - channel [RDPDR] received 21328, expected state RDPDR_CHANNEL_STATE_NAME_REQUEST but have state RDPDR_CHANNEL_STATE_INITIAL, aborting.
[14:55:33:516] [48256:48269] [ERROR][com.freerdp.channels.rdpdr.client] - channel [RDPDR] received 17219, expected state RDPDR_CHANNEL_STATE_CLIENT_CAPS but have state RDPDR_CHANNEL_STATE_INITIAL, aborting.

```

I tried different combinations of Linux/Windows client/server and here's the result:

freerdp + xrdp = Drive redirection doesn't work
mstsc + xrdp = Drive redirection works
freerdp + Windows RDP = Drive redirection works

I don't know who's fault is this: `freerdp` or `xrdp`. The only errors I see in xrdp logs are:

```
[20231023-15:03:34] [ERROR] dynamic_monitor_open_response: error
[20231023-15:03:34] [ERROR] xrdp_rdp_recv: xrdp_channel_process failed
```

OS: Alt Linux (both client and server).
Kernel (client): 5.10.197
Kernel (server): 5.10.188
FreeRDP: 2.11.2
xrdp: 0.9.23.1

**Comment by akallabeth** [comment]
@Mogaba I´ve opened a bug report with [xrdp](https://github.com/neutrinolabs/xrdp/issues/2834)
this looks like the server sends the drive initialization messages in a wrong order.

**Comment by matt335672** [comment]
@Mogaba - xrdp maintainer here.

Can you tell me which version of alt linux I should download to reproduce this? It's not clear to me from your post what I should use.

Thanks

**Reply by Mogaba** [reply]
Sorry for the late response, I use LXC container as xrdp server:
https://uk.lxd.images.canonical.com/images/alt/p10/amd64/default/20230830_01:17/

If you need a full OS you can use any image from here:
http://mirror.yandex.ru/altlinux/p10/images/

Also I did some testing and found out that older versions of FreeRDP (I tried 2.10.0) work OK so I guess xrdp is not the culprit here?

**Reply by akallabeth** [reply]
@Mogaba we´ve added some state checks for initialization (details see [[MS-RDPEFS]](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/006b6528-0684-438e-8c62-e30c8c27f0f0) ) and these hit a bug in `xrdp` here (it definitely sends a message in wrong order)

**Reply by matt335672** [reply]
Agreed

**Reply by akallabeth** [reply]
@matt335672 but be aware, looks like I´ve also still work to do, the checks in `FreeRDP` seem to be not 100% correct, see https://github.com/FreeRDP/FreeRDP/issues/9506
the handling of the `PAKID_CORE_USER_LOGGEDON` message has some rough edges that I need to iron out.

**Reply by matt335672** [reply]
Thanks, and interesting you say that.

My copy of [MS-RDPEFS] is v20210625 June 2021 which I believe is the latest. The message sequence diagram in 4.1 doesn't appear to be quite right - it shows this order:-

```
--------- Server Core Capability Request -------->
<-------- Client Core Capability Response --------
------------ Server Client ID Confirm ----------->
```

when I implemented this however, I wasn't seeing the client core capability response from the Windows 10 mstsc.exe I was using.

A careful re-reading of 3.3.5.1.7 suggested that the correct order should be this:-
```
--------- Server Core Capability Request -------->
------------ Server Client ID Confirm ----------->
<-------- Client Core Capability Response --------
```

In other words, it looks to me that the diagrams and the text may not necessarily correspond.

**Reply by akallabeth** [reply]
@matt335672 funny you mention 4.1, I modeled after 1.3.1 where it is exactly how you found it works. ah, consistency in specs is only for the faint harted :grin:

**Reply by matt335672** [reply]
In my time I've written my fair share of inconsistent specs which I'm sure have confused my successors too!

**Comment by matt335672** [comment]
I've implemented a fix in xrdp which I'm happy with, so I've closed the xrdp bug report for now. I'm pretty sure we're compliant with the spec. now.

@akallabeth - please come back to me if there are still issues when FreeRDP is updated.

**Reply by akallabeth** [reply]
@matt335672 your changes look good and are now like expected. now need to find time to fix it on my end.
